### ISSUE TITLE
Optimize for AnsiEsc plugin

### ISSUE THREAD

**Issue by jerryyin** [sainnhe/gruvbox-material#220]
I'm using `gruvbox-material` with `Plug 'powerman/vim-plugin-AnsiEsc'`. As it turns out AnsiEsc didn't use the color palette defined by gruvbox-material. 

### To reproduce:

> ls --color &> color
> vim color
> :AnsiEsc

The color displayed via AnsiEsc is using default vim defined color scheme, not from gruvbox. Taking a look at AnsiEsc source code, colors are defined like:

https://github.com/powerman/vim-plugin-AnsiEsc/blob/690f820d20b6e3a79ba20499874eb7333aa4ca5c/autoload/AnsiEsc.vim#L1823

The ctermfg are fairly standard. I'm wondering why it isn't picking up the colors defined from gruvbox material? In attempting to get it work, setting `gruvbox_material_custom()` doesn't seem to make any difference.

**Comment by antoineco** [comment]
I don't think those `ansiBlue`, etc. highlight groups are standard. The plugin links them to color names which are hardcoded into Vim, the it is expected that colors will be off.

If you override for example `ansiBlue` as below (from Vim's command line), then call your plugin again, do you see any change? 

```vim
" 'Blue' is a color defined by gruvbox-material
:hi! link ansiBlue Blue
```

**Comment by jerryyin** [comment]
@antoineco Thank you for the fast reply!

I've made a few attempts and none seem to make a difference:
 1. `:hi! link ansiBlue Blue`
 2. Define this in .vimrc
 3. Use `call gruvbox_material#highlight('ansiBlue', l:palette.aqua, l:palette.none)` in ` gruvbox_material_custom()`

**Comment by antoineco** [comment]
I just installed the plugin and opened one of its test files: [simple.ansi](https://github.com/powerman/vim-plugin-AnsiEsc/blob/b079ab1441a1cdb72abf462f8260fc46284b5dab/t/data/files/simple.ansi)

![](https://github.com/user-attachments/assets/a2e34aaf-7881-4ccb-914f-95b8a176bffe)

To provide some context about the issue you described, here is how the escapes are currently rendered:

![](https://github.com/user-attachments/assets/4271c7b9-2c4d-4c76-b9eb-0be9e4d888c7)

Inspecting the effective highlight group using Neovim's `:Inspect` command gives me this:

![](https://github.com/user-attachments/assets/5aeca32f-12db-4a5c-8dc8-5bdc36684c15)
![](https://github.com/user-attachments/assets/2085f5b2-14f6-4fd6-b1d5-8effc8eac349)

After issuing the Vim command `:hi! link ansiGreenFg Green`, the green text turns to "Gruvbox Material Green":

![](https://github.com/user-attachments/assets/77cd03bd-82cd-429d-ac0e-c7c93d963406)

Some highlight groups are more complex than just a foreground colors, for example:
- ![](https://github.com/user-attachments/assets/a2f6990b-81a9-46d2-8984-386e8e31b330)
- ![](https://github.com/user-attachments/assets/21d66211-512d-4dcc-aa47-f6fce19430b3)

In short, supporting this plugin is mostly about identifying the different combinations of ANSI escapes it supports, and the corresponding highlight groups. Feel free to try a few things on your side via `gruvbox_material_custom()` and then we can decide how to better optimize the colorscheme for this plugin.

**Comment by jerryyin** [comment]
Hello @antoineco, you've given me a nice tutorial about how to find out the color for the syntax group! Appreciate the effort in giving so much details in the full process about the issue. Turns out I didn't have the right syntax group name in previous attempt.

I've played around it a bit more after seeing your instructions and can confirm `:hi! link` for green (scripts) and blue (folders) will work after `AnsiEsc` will make highlighting groups picking up the right colors.

One caveat I've found out is that the workaround needs to happen **after** `:AnsiEsc` is invoked:
 - First approach, as you suggested `:hi! link ansiBlueFg Blue`
 - Second approach, put `call gruvbox_material#highlight('ansiBlueFg', l:palette.aqua, l:palette.none)` in `gruvbox_material_custom()`, and manually reload the colorscheme via `:colorscheme gruvbox-material`

Any effort trying to make the highlighting group work before `:AnsiEsc` invoked doesn't work. I've tried to put this either directly in the vimrc, or as customize highlighting group in `gruvbox_material_custom()`, they don't seem to make a difference unless I manually reload the color again.

**Comment by antoineco** [comment]
You're right, the plugin seems to override all highlight groups whenever it's loaded 😞 ([example](https://github.com/powerman/vim-plugin-AnsiEsc/blob/690f820d20b6e3a79ba20499874eb7333aa4ca5c/autoload/AnsiEsc.vim#L1551-L1559))

Two workarounds I can think of are:

1. Create an autocommand on the relevant event (I have to look through `:h events` and pick the right one), and re-apply highlight groups whenever the plugin gets loaded.
1. Submit a pull-request to the plugin's maintainer and suggest them to use the `default` highlight attribute:
    ```vimhelp
    				*:hi-default* *:highlight-default*
    The [default] argument is used for setting the default highlighting for a
    group.	If highlighting has already been specified for the group the command
    will be ignored.  Also when there is an existing link.
   ```

**Comment by jerryyin** [comment]
Thanks for the follow-up! 

Re your suggested workarounds:
 1. I think a `BufWinEnter` will do the job
 2. I tried a string replacement over the `AnsiEsc.vim` `%s/hi ansi/hi default ansi/g` but didn't see gruvbox loading up. It may need some additional tweaks for the plugin to work as expected (and looks like the plugin isn't been actively maintained for a few years, so chances that I can get a PR in isn't very high).

Luckily speaking, I'm not the first one to be complaining about the poor colorscheme of the ansiesc plugin. Turns out I can use https://github.com/Makaze/AnsiEsc and manually override the colors. The approach seems quite hacky but at least offers a solution that can work out-of-box.

I'm closing the issue since this isn't relevant to gruvbox-material. Thank you so much for your great work in this plugin.

**Comment by antoineco** [comment]
The `default` attribute only works if the highlight groups are already set when the plugin gets loaded. This might indeed not be the case if these highlight groups are set using an autocommand.

Glad you found that fork! It seems like a viable alternative.
